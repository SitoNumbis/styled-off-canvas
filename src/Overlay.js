import React, { useContext } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Context from './Context'

const Div = styled.div(({ color, opacity, duration, isOpen }) => `
  background: ${color || '#000'};
  height: 100%;
  left: 0;
  opacity: ${isOpen ? opacity || 0.2 : 0};
  position: fixed;
  top: 0;
  transition: opacity ${duration || '500ms'} ease-in-out;
  width: 100%;
  z-index: ${isOpen ? 1 : -1};
`)

const Overlay = (props) => {
  const { isOpen } = useContext(Context)

  return <Div {...props} isOpen={isOpen} />
}

Overlay.propTypes = {
  color: PropTypes.string,
  duration: PropTypes.string,
  opacity: PropTypes.number
}

export default Overlay
