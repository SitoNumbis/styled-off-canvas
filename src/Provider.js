import React from 'react'
import PropTypes from 'prop-types'

import Context from './Context'

/**
 * Provider component
 *
 * @param   {boolean}   closeOnEsc        if the menu should close on esc keydown
 * @param   {boolean}   isOpen            if the menu should be visible or not
 * @param   {string}    menuBackground    background color of the menu
 * @param   {string}    menuDuration      duration of the open/close animation for the menu
 * @param   {function}  onClose           callback function if menu closes
 * @param   {string}    overlayBackground background color of the menu
 * @param   {string}    overlayDuration   duration of the open/close animation for the overlay
 * @param   {string}    overlayOpacity    opacity of the overlay
 * @param   {string}    position          position of the menu (left or right)
 * @param   {string}    width             maximum width of the menu
 * @returns {component}                   React component
 */
const Provider = ({
  children,
  closeOnEsc = true,
  isOpen = false,
  menuBackground = '#fff',
  menuDuration = '500ms',
  onClose,
  overlayBackground = '#000',
  overlayDuration = '500ms',
  overlayOpacity = 0.2,
  position = 'right',
  width = '300px'
}) => (
  <Context.Provider
    value={{
      closeOnEsc,
      isOpen,
      menuBackground,
      menuDuration,
      onClose,
      overlayBackground,
      overlayDuration,
      overlayOpacity,
      position,
      width
    }}
  >
    {children}
  </Context.Provider>
)

Provider.propTypes = {
  closeOnEsc: PropTypes.bool,
  isOpen: PropTypes.bool,
  menuBackground: PropTypes.string,
  menuDuration: PropTypes.string,
  onClose: PropTypes.func.isRequired,
  overlayBackground: PropTypes.string,
  overlayDuration: PropTypes.string,
  overlayOpacity: PropTypes.number,
  position: PropTypes.oneOf(['left', 'right']),
  width: PropTypes.string
}

export default Provider
