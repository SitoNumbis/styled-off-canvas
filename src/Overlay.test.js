import React from 'react'
import renderer from './helper/tests/jest'

import Provider from './Provider'
import Overlay from './Overlay'

describe('Overlay', () => {
  test('default', () => {
    const tree = renderer(<Provider onClose={() => {}} isOpen><Overlay /></Provider>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('custom color', () => {
    const tree = renderer(<Provider onClose={() => {}} isOpen overlayBackground='ccc'><Overlay /></Provider>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('custom opacity', () => {
    const tree = renderer(<Provider onClose={() => {}} isOpen overlayOpacity={0.5}><Overlay /></Provider>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('custom duration', () => {
    const tree = renderer(<Provider onClose={() => {}} isOpen overlayDuration='1s'><Overlay /></Provider>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('closed', () => {
    const tree = renderer(<Provider onClose={() => {}}><Overlay /></Provider>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('close on click', () => {
    const cb = jest.fn()
    const component = renderer(<Provider onClose={cb}><Overlay /></Provider>)

    component.root.children[0].children[0].props.onClick()

    expect(cb).toHaveBeenCalled()
  })
})
