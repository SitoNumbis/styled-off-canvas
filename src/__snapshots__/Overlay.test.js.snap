// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay closed 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 bdfQkp"
  onClick={[Function]}
  opacity={0.2}
/>
`;

exports[`Overlay custom color 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 eiJRDc"
  onClick={[Function]}
  opacity={0.2}
/>
`;

exports[`Overlay custom duration 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 eiTvOF"
  onClick={[Function]}
  opacity={0.2}
/>
`;

exports[`Overlay custom opacity 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 jYUzeH"
  onClick={[Function]}
  opacity={0.5}
/>
`;

exports[`Overlay default 1`] = `
<div
  className="Overlay__Div-sc-7f8d3g-0 hCazpk"
  onClick={[Function]}
  opacity={0.2}
/>
`;
